api/client/trust.go

```diff
@@ -0,0 +1,525 @@
+		return err
+	}
+
+	if ref.String() == "" {
+		// List all targets
+		targets, err := notaryRepo.ListTargets()
+		if err != nil {
+			return err
+		}
+		for _, tgt := range targets {
+			t, err := convertTarget(*tgt)
+			if err != nil {
+				fmt.Fprintf(cli.out, "Skipping target for %q\n", repoInfo.LocalName)
+				continue
+			}
+			refs = append(refs, t)
+		}
+	} else {
+		t, err := notaryRepo.GetTargetByName(ref.String())
+		if err != nil {
+			return err
+		}
+		r, err := convertTarget(*t)
+		if err != nil {
+			return err
+
+		}
+		refs = append(refs, r)
+	}
+
+	v.Set("fromImage", repoInfo.LocalName)
+	for i, r := range refs {
+		displayTag := r.reference.String()
+		if displayTag != "" {
+			displayTag = ":" + displayTag
+		}
+		fmt.Fprintf(cli.out, "Pull (%d of %d): %s%s@%s\n", i+1, len(refs), repoInfo.LocalName, displayTag, r.digest)
+		v.Set("tag", r.digest.String())
+
+		_, _, err = cli.clientRequestAttemptLogin("POST", "/images/create?"+v.Encode(), nil, cli.out, repoInfo.Index, "pull")
+		if err != nil {
+			return err
+		}
+
+		// If reference is not trusted, tag by trusted reference
+		if !r.reference.IsTrusted() {
+			if err := cli.tagTrusted(repoInfo, registry.DigestReference(r.digest), r.reference); err != nil {
+				return err
+
+			}
+		}
+	}
+	return nil
+}
+
+func targetStream(in io.Writer) (io.WriteCloser, <-chan []target) {
+	// Create multiWriter
+	r, w := io.Pipe()
+	// Capture output
+	out := io.MultiWriter(in, w)
+	// output chan
+	targetChan := make(chan []target)
+
+	go func() {
+		targets := []target{}
+		scanner := bufio.NewScanner(r)
+		scanner.Split(ScanInput)
+		// New splitter
+		for scanner.Scan() {
+			line := scanner.Text()
+			if matches := targetRegexp.FindStringSubmatch(line); len(matches) == 4 {
+				dgst, err := digest.ParseDigest(matches[2])
+				if err != nil {
+					fmt.Fprintf(in, "Bad digest value %q, ignoring\n", matches[2])
+					continue
+				}
+				s, err := strconv.ParseInt(matches[3], 10, 64)
+				if err != nil {
+					fmt.Fprintf(in, "Bad size value %q, ignoring\n", matches[3])
+					continue
+				}
+
+				targets = append(targets, target{
+					reference: registry.ParseReference(matches[1]),
+					digest:    dgst,
+					size:      s,
+				})
+			}
+		}
+		targetChan <- targets
+	}()
+
+	type streamWriteCloser struct {
+		io.Writer
+		io.Closer
+	}
+	return streamWriteCloser{
+		Writer: out,
+		Closer: w,
+	}, targetChan
+}
+
+func (cli *DockerCli) trustedPush(repoInfo *registry.RepositoryInfo, tag string, authConfig cliconfig.AuthConfig) error {
+	streamOut, targetChan := targetStream(cli.out)
+
+	v := url.Values{}
+	v.Set("tag", tag)
+
+	_, _, err := cli.clientRequestAttemptLogin("POST", "/images/"+repoInfo.RemoteName+"/push?"+v.Encode(), nil, streamOut, repoInfo.Index, "push")
+	if err != nil {
+		return err
+	}
+
+	// Close stream channel to finish target parsing
+	if err := streamOut.Close(); err != nil {
```

Comment:
Hmm, guessing the scanner would hang since I don't think an EOF will be sent. Could always close before checking the previous error.